### PR TITLE
Better reconnection logic and better makeBuf

### PR DESCRIPTION
**Pull request by Syeberman**

### CODE REVIEW

File: le.go [hunk 1]
```diff
@@ -186,31 +170,66 @@ func (logger *Logger) SetPrefix(prefix string) {
...
-	logger.buf = append(logger.buf, p...)
-
-	if !strings.HasSuffix(string(logger.buf), lineSep) {
-		logger.buf = append(logger.buf, (lineSep)...)
+func (logger *Logger) makeBuf(p []byte) []byte {
+	// Pre-allocate a buffer of the correct size
+	capacity := len(logger.token) + 1
+	capacity += len(logger.prefix) + 1
+	capacity += len(p)                         // nominal payload size (before replacement)
+	capacity += countByte(p, asciiLineSep) * 2 // 1-byte "\n"s replaced with 3-byte "\u2028"s
```
Comment (MahlerFive) [inline le.go:207]: It looks like the original code replaced all but 1 line separator. Does `p` always end with a line sep which they were keeping? If so, it looks like you're adding 2 extra bytes on this line, and 1 extra on the next line.

File: le.go [hunk 2]
```diff
@@ -23,19 +21,23 @@ import (
...
+)
+
+var unicodeLineSep = []byte{0xE2, 0x80, 0xA8} // "\u2028"
 
 // Connect creates a new Logger instance and opens a TCP connection to
 // logentries.com,
 // The token can be generated at logentries.com by adding a new log,
 // choosing manual configuration and token based TCP connection.
 func Connect(token string) (*Logger, error) {
+	fmt.Println("le_go: Connect()")
```
Comment (MahlerFive) [inline le.go:40]: If this is just for running a test and helping us debug then I'm fine with all the debug prints for now, but we'll need to remove all of this after and make it acceptable to make a PR against the original repo.

File: le.go [hunk 3]
```diff
@@ -186,31 +170,66 @@ func (logger *Logger) SetPrefix(prefix string) {
...
 
-	logger.makeBuf(p)
+	n, err := logger.conn.Write(buf)
+	if err == nil {
+		return n, err
+	}
+	fmt.Printf("le_go: Write n=%d, err=%s", n, err.Error())
 
-	return logger.conn.Write(logger.buf)
+	// First write failed.  Try reconnecting and then a second write; if that fails give up.  If
```
Comment (MahlerFive) [inline le.go:184]: I think it doesn't make sense to ever keep retrying to connect during a write(), just let the caller handle that how they want when they get an error back.

File: le.go [hunk 4]
```diff
@@ -186,31 +170,66 @@ func (logger *Logger) SetPrefix(prefix string) {
...
+			break
+		}
+
+		buf = append(buf, p[:i]...)
+		buf = append(buf, unicodeLineSep...)
+		p = p[i+1:]
 	}
+
+	// Buffer must end with an ascii line separator
+	buf = append(buf, asciiLineSep)
```
Comment (MahlerFive) [inline le.go:232]: Does `p` already end in asciiLineSep? If so we don't need to add another.

**Review by MahlerFive — CHANGES_REQUESTED** [review]

File: le.go (hunk 2 above)
Comment (Syeberman) [reply le.go:40]: Yeah

File: le.go (hunk 3 above)
Comment (Syeberman) [reply le.go:184]: But these writes are triggered by logger prints scattered everywhere in the code.  Imagine every time you wanted to print some information, you had to attempt a reconnection.

File: le.go (hunk 1 above)
Comment (Syeberman) [reply le.go:207]: Yeah, it did, but the Python version of this client doesn't bother: it replaces all newlines, then adds a `\n` at the end, unconditionally.

File: le.go (hunk 4 above)
Comment (Syeberman) [reply le.go:232]: Yeah, but the Python version doesn't bother.

File: le.go (hunk 3 above)
Comment (MahlerFive) [reply le.go:184]: True - I guess your comment makes sense here then. In either case, shouldn't we be passing the errors back up to the caller? Right now calling `Println()` won't return the error that gets returned from `Write()`.

**Review by MahlerFive — CHANGES_REQUESTED** [review]

File: le.go (hunk 3 above)
Comment (Syeberman) [reply le.go:184]: I can't change an existing function interface.  All downstream code would break.

**Review by MahlerFive — APPROVED** [review]

**Pull request merged**